File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/jobs/TraktEpisodeJob.java
package com.battlelancer.seriesguide.jobs;

import android.content.Context;
import android.support.annotation.NonNull;
import com.battlelancer.seriesguide.SgApp;
import com.battlelancer.seriesguide.enums.EpisodeFlags;
import com.battlelancer.seriesguide.jobs.episodes.JobAction;
import com.battlelancer.seriesguide.settings.TraktCredentials;
import com.battlelancer.seriesguide.traktapi.SgTrakt;
import com.battlelancer.seriesguide.util.EpisodeTools;
import com.battlelancer.seriesguide.util.ShowTools;
import com.uwetrottmann.trakt5.entities.ShowIds;
import com.uwetrottmann.trakt5.entities.SyncEpisode;
import com.uwetrottmann.trakt5.entities.SyncItems;
import com.uwetrottmann.trakt5.entities.SyncResponse;
import com.uwetrottmann.trakt5.entities.SyncSeason;
import com.uwetrottmann.trakt5.entities.SyncShow;
import com.uwetrottmann.trakt5.services.Sync;
import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import org.threeten.bp.Instant;
import org.threeten.bp.OffsetDateTime;
import org.threeten.bp.ZoneOffset;
import retrofit2.Call;
import retrofit2.Response;

public class TraktEpisodeJob extends NetworkJob {

    private final long actionAtMs;
    private Integer showTraktId;

    public TraktEpisodeJob(JobAction action, SgJobInfo jobInfo, long actionAtMs) {
        super(action, jobInfo);
        this.actionAtMs = actionAtMs;
    }

    public boolean checkCanUpload(Context context) {
        // Do not send if show has no trakt id (was not on trakt last time we checked).
        showTraktId = ShowTools.getShowTraktId(context, jobInfo.showTvdbId());
        return showTraktId != null;
    }

    public int upload(Context context) {
        final int flagValue = jobInfo.flagValue();

        // can not (yet) track with trakt or is skipped flag not supported by trakt
        if (showTraktId == null || EpisodeTools.isSkipped(flagValue)) {
            return NetworkJob.SUCCESS;
        }

        boolean isAddNotDelete = flagValue
                != EpisodeFlags.UNWATCHED; // 0 for not watched or not collected
        List<SyncSeason> seasons = getEpisodesForTrakt(isAddNotDelete);
        if (seasons.isEmpty()) {
            return NetworkJob.SUCCESS; // nothing to upload, done.
        }

        if (!TraktCredentials.get(context).hasCredentials()) {
            return NetworkJob.ERROR_TRAKT_AUTH;
        }

        // outer wrapper and show are always required
        SyncShow show = new SyncShow().id(ShowIds.trakt(showTraktId));
        SyncItems items = new SyncItems().shows(show);
        show.seasons(seasons);

        // determine network call
        String errorLabel;
        Call<SyncResponse> call;
        Sync traktSync = SgApp.getServicesComponent(context).traktSync();
        switch (action) {
            case EPISODE_WATCHED_FLAG:
                if (isAddNotDelete) {
                    errorLabel = "set episodes watched";
                    call = traktSync.addItemsToWatchedHistory(items);
                } else {
                    errorLabel = "set episodes not watched";
                    call = traktSync.deleteItemsFromWatchedHistory(items);
                }
                break;
            case EPISODE_COLLECTION:
                if (isAddNotDelete) {
                    errorLabel = "add episodes to collection";
                    call = traktSync.addItemsToCollection(items);
                } else {
                    errorLabel = "remove episodes from collection";
                    call = traktSync.deleteItemsFromCollection(items);
                }
                break;
            default:
                throw new IllegalArgumentException("Action " + action + " not supported.");
        }

        // execute call
        try {
            Response<SyncResponse> response = call.execute();
            if (response.isSuccessful()) {
                // check if any items were not found
                if (isSyncSuccessful(response.body())) {
                    return NetworkJob.SUCCESS;
                }
            } else {
                if (SgTrakt.isUnauthorized(context, response)) {
                    return NetworkJob.ERROR_TRAKT_AUTH;
                }
                SgTrakt.trackFailedRequest(context, errorLabel, response);
            }
        } catch (IOException e) {
            SgTrakt.trackFailedRequest(context, errorLabel, e);
        }
        return NetworkJob.ERROR_TRAKT_API;
    }

    /**
     * Builds a list of {@link com.uwetrottmann.trakt5.entities.SyncSeason} objects to submit to
     * trakt.
     */
    @NonNull
    private List<SyncSeason> getEpisodesForTrakt(boolean isAddNotDelete) {
        // send time of action to avoid adding duplicate plays/collection events at trakt
        // if this job re-runs due to failure, but trakt already applied changes (it happens)
        // also if execution is delayed to due being offline this will ensure
        // the actual action time is stored at trakt
        Instant instant = Instant.ofEpochMilli(actionAtMs);
        OffsetDateTime actionAtDateTime = instant.atOffset(ZoneOffset.UTC);

        List<SyncSeason> seasons = new ArrayList<>();

        SyncSeason currentSeason = null;
        for (int i = 0; i < jobInfo.episodesLength(); i++) {
            EpisodeInfo episodeInfo = jobInfo.episodes(i);

            int seasonNumber = episodeInfo.season();

            // start new season?
            if (currentSeason == null || seasonNumber > currentSeason.number) {
                currentSeason = new SyncSeason().number(seasonNumber);
                currentSeason.episodes = new LinkedList<>();
                seasons.add(currentSeason);
            }

            // add episode
            SyncEpisode episode = new SyncEpisode().number(episodeInfo.number());
            if (isAddNotDelete) {
                // only send timestamp if adding, not if removing to save data
                if (action == JobAction.EPISODE_WATCHED_FLAG) {
                    episode.watchedAt(actionAtDateTime);
                } else {
                    episode.collectedAt(actionAtDateTime);
                }
            }
            currentSeason.episodes.add(episode);
        }

        return seasons;
    }

    /**
     * If the {@link SyncResponse} is invalid or any show, season or episode was not found
     * returns {@code false}.
     */
    private static boolean isSyncSuccessful(SyncResponse response) {
        if (response == null || response.not_found == null) {
            // invalid response, assume failure
            return false;
        }

        if (response.not_found.shows != null && !response.not_found.shows.isEmpty()) {
            // show not found
            return false;
        }
        if (response.not_found.seasons != null && !response.not_found.seasons.isEmpty()) {
            // show exists, but seasons not found
            return false;
        }
        //noinspection RedundantIfStatement
        if (response.not_found.episodes != null && !response.not_found.episodes.isEmpty()) {
            // show and season exists, but episodes not found
            return false;
        }

        return true;
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/jobs/TraktEpisodeJob.kt
package com.battlelancer.seriesguide.jobs

import android.content.Context
import com.battlelancer.seriesguide.SgApp
import com.battlelancer.seriesguide.jobs.episodes.JobAction
import com.battlelancer.seriesguide.shows.episodes.EpisodeFlags
import com.battlelancer.seriesguide.shows.episodes.EpisodeTools
import com.battlelancer.seriesguide.traktapi.SgTrakt
import com.battlelancer.seriesguide.traktapi.TraktCredentials
import com.battlelancer.seriesguide.util.Errors
import com.github.michaelbull.result.Err
import com.github.michaelbull.result.Ok
import com.github.michaelbull.result.Result
import com.github.michaelbull.result.getOrElse
import com.uwetrottmann.trakt5.TraktV2
import com.uwetrottmann.trakt5.entities.ShowIds
import com.uwetrottmann.trakt5.entities.SyncEpisode
import com.uwetrottmann.trakt5.entities.SyncItems
import com.uwetrottmann.trakt5.entities.SyncResponse
import com.uwetrottmann.trakt5.entities.SyncSeason
import com.uwetrottmann.trakt5.entities.SyncShow
import com.uwetrottmann.trakt5.entities.UserSlug
import com.uwetrottmann.trakt5.enums.HistoryType
import com.uwetrottmann.trakt5.services.Users
import org.threeten.bp.Instant
import org.threeten.bp.OffsetDateTime
import org.threeten.bp.ZoneOffset
import retrofit2.Call
import java.util.LinkedList

class TraktEpisodeJob(
    action: JobAction,
    jobInfo: SgJobInfo,
    private val actionAtMs: Long
) : BaseNetworkEpisodeJob(action, jobInfo) {

    override fun execute(context: Context): NetworkJobResult {
        // Do not send if show has no trakt id (was not on trakt last time we checked).
        val showTraktId = SgApp.getServicesComponent(context)
            .showTools().getShowTraktId(jobInfo.showId())
        val canSendToTrakt = showTraktId != null
        if (!canSendToTrakt) {
            return buildResult(context, ERROR_TRAKT_NOT_FOUND)
        }
        val result = upload(context, showTraktId!!)
        return buildResult(context, result)
    }

    private fun upload(context: Context, showTraktId: Int): Int {
        val flagValue = jobInfo.flagValue()

        // skipped flag not supported by trakt
        if (EpisodeTools.isSkipped(flagValue)) {
            return SUCCESS
        }

        val isAddNotDelete = (flagValue
                != EpisodeFlags.UNWATCHED) // 0 for not watched or not collected
        val seasons = getEpisodesForTrakt(context, showTraktId, isAddNotDelete)
            .getOrElse { return it }
        if (seasons.isEmpty()) {
            return SUCCESS // nothing to upload, done.
        }

        if (!TraktCredentials.get(context).hasCredentials()) {
            return ERROR_TRAKT_AUTH
        }

        // outer wrapper and show are always required
        val show = SyncShow().id(ShowIds.trakt(showTraktId))
        val items = SyncItems().shows(show)
        show.seasons(seasons)

        // determine network call
        val errorLabel: String
        val call: Call<SyncResponse>

        val trakt = SgApp.getServicesComponent(context).trakt()
        val traktSync = trakt.sync()
        when (action) {
            JobAction.EPISODE_WATCHED_FLAG -> if (isAddNotDelete) {
                errorLabel = "set episodes watched"
                call = traktSync.addItemsToWatchedHistory(items)
            } else {
                errorLabel = "set episodes not watched"
                call = traktSync.deleteItemsFromWatchedHistory(items)
            }
            JobAction.EPISODE_COLLECTION -> if (isAddNotDelete) {
                errorLabel = "add episodes to collection"
                call = traktSync.addItemsToCollection(items)
            } else {
                errorLabel = "remove episodes from collection"
                call = traktSync.deleteItemsFromCollection(items)
            }
            else -> throw IllegalArgumentException("Action $action not supported.")
        }

        // execute call
        try {
            val response = call.execute()
            if (response.isSuccessful) {
                // check if any items were not found
                if (!isSyncSuccessful(response.body())) {
                    return ERROR_TRAKT_NOT_FOUND
                }
            } else {
                if (SgTrakt.isUnauthorized(context, response)) {
                    return ERROR_TRAKT_AUTH
                }
                Errors.logAndReport(
                    errorLabel, response,
                    SgTrakt.checkForTraktError(trakt, response)
                )
                val code = response.code()
                return if (code == 429 /* Rate Limit Exceeded */ || code >= 500) {
                    ERROR_TRAKT_SERVER
                } else {
                    ERROR_TRAKT_CLIENT
                }
            }
        } catch (e: Exception) {
            Errors.logAndReport(errorLabel, e)
            return ERROR_CONNECTION
        }
        return SUCCESS
    }

    /**
     * Builds a list of [SyncSeason] objects to submit to Trakt. When adding watched history
     * entries, checks against Trakt history if an entry exists at that time for an episode
     * and excludes those episodes. This is to prevent duplicate entries due to sending failing
     * due to a network error, but changes getting still applied at Trakt.
     */
    private fun getEpisodesForTrakt(
        context: Context,
        showTraktId: Int,
        isAddNotDelete: Boolean
    ): Result<List<SyncSeason>, Int> {
        val isAddingWatchedEntry = action == JobAction.EPISODE_WATCHED_FLAG && isAddNotDelete

        // Send time of action to avoid adding duplicate collection entries at Trakt (does not work
        // for watched entries, separate check for those below) if this job re-runs due to failure,
        // but Trakt already applied changes (it happens).
        // Also if execution is delayed due to being offline this will ensure the actual action time
        // is stored at Trakt.
        val instant = Instant.ofEpochMilli(actionAtMs)
        val actionAtDateTime = instant.atOffset(ZoneOffset.UTC)

        val seasons: MutableList<SyncSeason> = ArrayList()
        val trakt = SgApp.getServicesComponent(context).trakt()
        val traktUsers = trakt.users()

        // If sending a watched entry, check if there are already episodes watched at
        // actionAtDateTime, then exclude those. This is to prevent duplicate watched entries at
        // Trakt if this job re-runs due to failure, but Trakt already applied changes.
        val watchedEpisodes = mutableListOf<WatchedEpisode>()
        if (isAddingWatchedEntry) {
            var page = 1
            do {
                val historyPage =
                    getHistoryEntryPage(
                        context,
                        trakt,
                        traktUsers,
                        showTraktId,
                        actionAtDateTime,
                        page
                    ).getOrElse { return Err(it) }
                watchedEpisodes.addAll(historyPage.episodes)
                page++
            } while (page <= historyPage.pageCount)
        }

        var currentSeason: SyncSeason? = null
        for (i in 0 until jobInfo.episodesLength()) {
            val episodeInfo = jobInfo.episodes(i)
            val number = episodeInfo.number()
            val seasonNumber = episodeInfo.season()

            if (isAddingWatchedEntry
                && watchedEpisodes.find { it.number == number && it.season == seasonNumber } != null) {
                // Skip, this episode already has an entry at actionAtDateTime.
                continue
            }

            // start new season?
            if (currentSeason == null || currentSeason.number?.let { seasonNumber > it } == true) {
                currentSeason = SyncSeason().number(seasonNumber)
                currentSeason.episodes = LinkedList()
                seasons.add(currentSeason)
            }

            // add episode
            val episode = SyncEpisode().number(episodeInfo.number())
            if (isAddNotDelete) {
                // only send timestamp if adding, not if removing to save data
                if (action == JobAction.EPISODE_WATCHED_FLAG) {
                    episode.watchedAt(actionAtDateTime)
                } else {
                    episode.collectedAt(actionAtDateTime)
                }
            }
            currentSeason.episodes!!.add(episode)
        }
        return Ok(seasons)
    }

    data class WatchedEpisode(
        val number: Int,
        val season: Int
    )

    data class HistoryPage(
        val episodes: List<WatchedEpisode>,
        val pageCount: Int
    )

    private fun getHistoryEntryPage(
        context: Context,
        trakt: TraktV2,
        traktUsers: Users,
        showTraktId: Int,
        actionAtDateTime: OffsetDateTime,
        page: Int
    ): Result<HistoryPage, Int> {
        val action = "get history of show"
        val historyCall = traktUsers.history(
            UserSlug.ME,
            HistoryType.SHOWS,
            showTraktId,
            page,
            null,
            null,
            actionAtDateTime,
            actionAtDateTime
        )
        return executeTraktCall(
            context,
            trakt,
            historyCall,
            action
        ) { response, body ->
            val episodes = body.map {
                val number = it.episode?.number
                val season = it.episode?.season
                if (number == null || season == null) {
                    Errors.logAndReport(action, response, "episode is null")
                    return@executeTraktCall Err(ERROR_TRAKT_CLIENT)
                }
                WatchedEpisode(number, season)
            }
            val pageCount = response.headers()["x-pagination-page-count"]?.toIntOrNull()
                ?: 1
            Ok(HistoryPage(episodes, pageCount))
        }
    }

    companion object {
        /**
         * If the [SyncResponse.not_found] indicates any show,
         * season or episode was not found returns `false`.
         */
        private fun isSyncSuccessful(response: SyncResponse?): Boolean {
            val notFound = response?.not_found ?: return true
            if (notFound.shows?.isNotEmpty() == true) {
                // show not found
                return false
            }
            if (notFound.seasons?.isNotEmpty() == true) {
                // show exists, but seasons not found
                return false
            }
            if (notFound.episodes?.isNotEmpty() == true) {
                // show and season exists, but episodes not found
                return false
            }
            return true
        }
    }
}